### PR TITLE
Esp wifi fix busy error after disconnect

### PR DESCRIPTION
**Pull request by KarolSzpila**
sock->tx_pkt is assigned in esp_sendto(). This packet has to be freed
eventually. In current implementation this is a responsibility of
_sock_send() which is called either directly, or indirectly by work
queue esp_send_work(). _sock_send() is freeing sock->tx_pkt in all
cases except when if (!esp_flag_is_set(dev, EDF_STA_CONNECTED)).

ESP modem is not notifying about closed UDP link when connection to
network is lost. This means that  sock->flag has still
ESP_SOCK_CONNETED bit set. On next esp_sendto() call, sock->tx_pkt
is assigned, then _sock_send() is called returning -ENETUNREACH
without freeing sock->tx_pkt. This results in permanent -EBUSY
error on every next esp_sendto() call.

This does not happen for TCP link because modem notifies about closed
TCP link. sock->flag ESP_SOCK_CONNETED bit is unset. In that case
next call esp_sendto() fails on esp_connect() returning -ENETUNREACH,
before sock->tx_pkt is assigned.

To solve this problem sock->tx_pkt must be also freed when there is no
WiFi connection.

### CODE REVIEW

**Review by mniestroj — DISMISSED** [review]
LGTM

**Comment by tsvehagen** [comment]
Thanks for the PR. It looks good but it made me see another bigger problem.

In `context_sendto()` in `net_context.c`, if the actual send call (in this case `net_offload_send`) fails, the packet is unreferenced by `context_sendto()` itself. So actually `net_pkt_unref` should not be called in the driver if `ret != 0`. It gets a bit more complicated though since _if_ the call was made with `timeout == 0`, `net_pkt_unref` should always be called since the `net_offload_send()` has already returned and `_sock_send()` is called from the worker thread.

I think maybe `net_pkt_unref` should be moved altogether from `_sock_send` and instead
- Always call `net_pkt_unref` in `esp_send_work`
- Call `net_pkt_unref` in `esp_sendto` _if_ `_sock_send` returned not 0

Someone please double check my logic :p

**Comment by mniestroj** [comment]
@tsvehagen I can confirm the problem with `net_pkt_unref` related to `context_sendto()`. I also agree with your proposal on how to fix it. However I won't 100% like the resulting code, as the logic about unreferencing tx_pkt will be hard to follow. That being said, I don't have better idea for now. It can be always improved further in future :)

**Comment by jukkar** [comment]
The logic in packet sending is that if the packet was sent properly, the driver will own it and is suppose to unref the successfully sent packet, but if there was an error, then the caller (in this case `context_sendto()`) will still own it and should free it. The original idea was that in error case, the sender could try to do something with the packet, like resending etc., but current code does not do implement this atm.

**Comment by tsvehagen** [comment]
@jukkar Right, that makes sense.

@KarolSzpila Do you want to update this PR to address what has been commented?

**Comment by KarolSzpila** [comment]
@tsvehagen, I agree with @mniestroj, that this would not be easy to follow solution. For now I would like to leave this PR as is.

I have noticed another thing. `timeout` argument is passed to esp_sendto(), but it is used only to call `_sock_send()` directly or from work. Is it really necessary to call
```
	if (timeout == 0) {
		k_work_submit_to_queue(&dev->workq, &sock->send_work);
		return 0;
	}
```
instead of call `_sock_send()` only directly. Then `net_pkt_unref` could be moved to `esp_sendto()`. Also I dont l

However, I think that this problem is more complicated and should be addressed in another PR.

**Comment by mniestroj** [comment]
@KarolSzpila Solution proposed by @tsvehagen will solve two issues at the same time - stop sock->tx_pkt leakage (addressed in your patches) and wrong handling of net_pkt depending on `enum net_verdict` returned. If we are about to apply current code, then we need to basically revert it in order to fix code a wider context.

So I think we should implement @tsvehagen proposal for now unless someone has better idea.

**Comment by tsvehagen** [comment]
> @tsvehagen, I agree with @mniestroj, that this would not be easy to follow solution. For now I would like to leave this PR as is.

I don't really agree with that solution being so hard to follow and even if it was it's still a bug that needs to be fixed.

> I have noticed another thing. `timeout` argument is passed to `esp_sendto()`, but it is used only to call `_sock_send()` directly or from work. Is it really necessary to call
> 
> ```
> 	if (timeout == 0) {
> 		k_work_submit_to_queue(&dev->workq, &sock->send_work);
> 		return 0;
> 	}
> ```
> 
> instead of call `_sock_send()` only directly. Then `net_pkt_unref` could be moved to `esp_sendto()`.

Yea that is done in order to get a non-blocking behaviour.

> However, I think that this problem is more complicated and should be addressed in another PR.

In what way is it more complicated?

Here is a PR that fixes both issues https://github.com/zephyrproject-rtos/zephyr/pull/26526.

**Comment by KarolSzpila** [comment]
> In what way is it more complicated?

I was talking about what @jukkar mentioned:

> The original idea was that in error case, the sender could try to do something with the packet, like resending etc., but current code does not do implement this atm.

@tsvehagen, thanks for PR. I thinks this PR could be closed since You resolved all issues.

**Comment by tsvehagen** [comment]
> > In what way is it more complicated?
> 
> I was talking about what @jukkar mentioned:
> 
> > The original idea was that in error case, the sender could try to do something with the packet, like resending etc., but current code does not do implement this atm.

Oh I see, yea that's another story.

> @tsvehagen, thanks for PR. I thinks this PR could be closed since You resolved all issues.

Okay, thanks for the help :+1:
